### PR TITLE
887 Update sharding limit

### LINKED ISSUE #887: Update sharding limit
**Author the user story for this feature**

As an Android engineer, I want to run as many shards as possible so I can execute test quickly.

**Describe the solution you'd like**

Let's update Flank to support the new 250 shard limit for virtual devices. The 50 shard limit for real devices remains in place. Let's also verify these limits work correctly by testing against the live FTL server.

**Additional context**
https://firebase-community.slack.com/archives/C72V6UW8M/p1594438861446800

### PR DESCRIPTION
**Pull request by adamfilipow92**
Fixes #887

For virtual devices, the limit should be set to 250.
For physical devices, the limit should be still 50.

In case when we have a virtual and physical device and shards limit set over the physical device limit, firebase returns an exception

```
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
{
  "code" : 400,
  "errors" : [ {
    "domain" : "global",
    "message" : "Given number of shards of 60 shards exceeds limit of 50 shards.",
    "reason" : "badRequest"


```

## Solution

If the limit exceeds the physical limit and we have physical and virtual devices we split matrices by device type. This solution allows us to use the maximum limit on every device type.

## Checklist

- [X] Documented
- [X] Unit tested
- [X] release_notes.md updated
- [X] Split devices by type and run on different matrices

### CODE REVIEW

File: test_runner/src/main/kotlin/ftl/run/NewTestRun.kt [hunk 1]
```diff
@@ -15,18 +15,17 @@ import ftl.util.FlankTimeoutError
...
-            cancelTestsOnTimeout(args.project, matrixMap.map) { fetchArtifacts(matrixMap, args) }
+    if (!args.async) {
+        cancelTestsOnTimeout(args.project, matrixMap.map) { pollMatrices(matrixMap.map.keys, args).update(matrixMap) }
+        cancelTestsOnTimeout(args.project, matrixMap.map) { fetchArtifacts(matrixMap, args) }
 
-            ReportManager.generate(matrixMap, args, testShardChunks, ignoredTests)
-        }
+        ReportManager.generate(matrixMap, args, testShardChunks, ignoredTests)
     }
+    return@withTimeoutOrNull this
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/run/NewTestRun.kt:28]: Is there any reason for returning CoroutineScope?

File: test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt [hunk 2]
```diff
@@ -29,36 +31,37 @@ internal suspend fun runAndroidTests(args: AndroidArgs): TestResult = coroutineS
...
-    val androidDeviceList = GcAndroidDevice.build(args.devices)
-
     val testMatrices = mutableListOf<Deferred<TestMatrix>>()
     val allTestShardChunks = mutableListOf<List<String>>()
     val ignoredTestsShardChunks = mutableListOf<List<String>>()
 
     val history = GcToolResults.createToolResultsHistory(args)
     val otherGcsFiles = args.uploadOtherFiles(runGcsPath)
     val additionalApks = args.uploadAdditionalApks(runGcsPath)
+    val argsList = if (args.shouldSplitRuns()) args.splitConfigurationByDeviceType() else listOf(args)
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt:41]: `argsList` is a little bit generic in this context maybe `deviceConfigurations` or `deviceArgs` will fits better?

File: test_runner/src/main/kotlin/ftl/run/NewTestRun.kt (hunk 1 above)
Comment (adamfilipow92) [reply test_runner/src/main/kotlin/ftl/run/NewTestRun.kt:28]: No, this is redundant I should remove it. Thanks!

File: test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt (hunk 2 above)
Comment (adamfilipow92) [reply test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt:41]: `deviceArgs` fits great, thanks!

File: test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt [hunk 3]
```diff
@@ -23,6 +25,7 @@ data class AndroidArgs(
 ) : IArgs by commonArgs {
     companion object : AndroidArgsCompanion()
 
+    // override val maxTestShards: Int = calculateMaxTestShards(commonArgs.maxTestShards)
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:28]: delete if unused

File: test_runner/src/main/kotlin/ftl/args/CreateIosArgs.kt [hunk 4]
```diff
@@ -17,9 +17,14 @@ private fun createIosArgs(
...
 ) = IosArgs(
-    commonArgs = commonArgs,
+    commonArgs = commonArgs.copy(maxTestShards = convertToShardCount(commonArgs.maxTestShards)),
     xctestrunZip = gcloud.test!!.processFilePath("from test"),
     xctestrunFile = gcloud.xctestrunFile!!.processFilePath("from xctestrun-file"),
     xcodeVersion = gcloud.xcodeVersion,
     testTargets = flank.testTargets!!.filterNotNull()
 )
+
+private fun convertToShardCount(inputValue: Int): Int =
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/args/CreateIosArgs.kt:27]: make single line, skip `: Int`

File: test_runner/src/main/kotlin/ftl/args/ValidateAndroidArgs.kt [hunk 5]
```diff
@@ -68,8 +69,28 @@ private fun AndroidArgs.assertDirectoriesToPull() {
...
+private fun AndroidArgs.assertDevicesShards() {
+    if (inVirtualRange && !inPhysicalRange) println("Physical devices configured, but max-test-shards limit set to $maxTestShards, for physical devices range is ${IArgs.AVAILABLE_PHYSICAL_SHARD_COUNT_RANGE.first} to ${IArgs.AVAILABLE_PHYSICAL_SHARD_COUNT_RANGE.last}, you additionally have configured virtual devices. In this case, the physical limit will be decreased to: ${IArgs.AVAILABLE_PHYSICAL_SHARD_COUNT_RANGE.last}")
+    else if (!inVirtualRange && !inPhysicalRange) throwMaxTestShardsLimitExceeded()
+}
+
+private fun AndroidArgs.assertVirtualDevicesShards() {
+    if (!inVirtualRange) throwMaxTestShardsLimitExceeded()
+}
+
+private fun AndroidArgs.throwMaxTestShardsLimitExceeded() {
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/args/ValidateAndroidArgs.kt:92]: return type could be `Nothing` for self documentation

File: test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt [hunk 6]
```diff
@@ -83,3 +86,17 @@ val AndroidArgs.isInstrumentationTest
 val AndroidArgs.isRoboTest
     get() = appApk != null &&
             (roboDirectives.isNotEmpty() || roboScript != null)
+
+fun AndroidArgs.containsVirtualDevices() = devices.any { it.isVirtual!! }
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:90]: `it.isVirtual ?: false` is safer call. WDYT?

File: test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt [hunk 7]
```diff
@@ -83,3 +86,17 @@ val AndroidArgs.isInstrumentationTest
...
+
+fun AndroidArgs.containsPhysicalDevices() = devices.any { !it.isVirtual!! }
+
+fun AndroidArgs.shouldSplitRuns() = containsPhysicalDevices() && maxTestShards > 50
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:94]: should not we use `AVAILABLE_PHYSICAL_SHARD_COUNT_RANGE.last` ?

File: test_runner/src/test/kotlin/ftl/fixtures/test_app_cases/flank-single-device-limit-success.yml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+      locale: en
+      orientation: portrait
+    - model: Nexus9
+      version: 25
+      locale: en
+      orientation: portrait
+flank:
+  disable-sharding: false
+  max-test-shards: 40
+
```
Comment (piotradamczyk5) [inline test_runner/src/test/kotlin/ftl/fixtures/test_app_cases/flank-single-device-limit-success.yml:21]: 1 empty line could be skipped ;)

File: test_runner/src/test/kotlin/ftl/fixtures/test_app_cases/flank-single-multi-device-fail.yml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+      locale: en
+      orientation: portrait
+    - model: Nexus9
+      version: 25
+      locale: en
+      orientation: portrait
+flank:
+  disable-sharding: false
+  max-test-shards: 251
+
```
Comment (piotradamczyk5) [inline test_runner/src/test/kotlin/ftl/fixtures/test_app_cases/flank-single-multi-device-fail.yml:21]: same here :)

File: test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt [hunk 10]
```diff
@@ -1,6 +1,8 @@
 package ftl.args
 
 import ftl.args.yml.AppTestPair
+import ftl.config.resolve
+import ftl.config.isVirtual
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/args/AndroidArgs.kt:5]: I think that these imports are unused

File: test_runner/src/main/kotlin/ftl/config/Device.kt [hunk 11]
```diff
@@ -12,9 +12,9 @@ data class Device(
     val model: String,
     val version: String,
     val locale: String = defaultLocale,
-    val orientation: String = defaultOrientation
+    val orientation: String = defaultOrientation,
+    val isVirtual: Boolean? = null
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/config/Device.kt:16]: `isVirtual` shouldn't be nullable. We don't need nulls in later calculations.

Comment (adamfilipow92) [reply test_runner/src/main/kotlin/ftl/config/Device.kt]: Alright! Changed! 👍

**Pull request merged**